rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,219 @@
+
+1. Even more effortless to set up.
+1. We've done this before for the embedded mode, [_remember_](https://github.com/WordPress/openverse-frontend/pull/65)?
+1. Testing is easy because tests can append the right flag param to the URL.
+
+##### Cons
+
+1. Configuration will need to live in the codebase so change of state will need deployment.
+1. Exposes half-baked features to users who know the flag.
+1. Transient, so flag must be set on every URL or stored in `localStorage` or cookie.
+1. Flags will be primitive and will not allow complex data. Even stringified JSON would be bad in URLs.
+
+##### Considerations
+
+We can clearly demarcate feature flag params with a prefix such as `ff_` to separate them from the actual query params and process them separately.
+
+Due to their nature, they are best used in pairing with another storage as an override switch rather than on their own.
+
+#### 3. Growthbook
+
+GrowthBook is an open-source platform for feature flagging. It provides SDKs and a UI to manage feature flags.
+
+##### Pros
+
+1. High-level solution with UI.
+1. Supports features we'd want later like A/B testing.
+1. Can record a large number of metrics.
+1. Actively maintained and FOSS.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes.
+
+##### Cons
+
+1. Adds API calls to the server startup.
+1. Separate service that needs to be self-hosted and set up.
+1. Will need Dockerisation of the dev workflow.
+1. Might be overkill if we don't want the extra features.
+1. We might not even want to use the metrics.
+
+##### Reference
+
+- [Home](https://www.growthbook.io)
+- [Docs](https://docs.growthbook.io)
+
+#### 4. API
+
+Our API could serve as a decent source of feature flag statuses by storing them as Django model.
+
+##### Pros
+
+1. Can be managed using Django admin.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes, essentially models.
+1. The frontend depends on the API anyway, so just add more of it.
+
+##### Cons
+
+1. Need to mock the API calls in dev environments.
+1. Adds coupling between frontend features and the API.
+
+#### 5. Conf files
+
+Conf files could be JSON/YAML/`.env` files that list the names of features and their values. They can live separate
+from the codebase.
+
+##### Pros
+
+1. Simple configuration.
+1. Prevents lots of API calls.
+
+##### Cons
+
+1. Lots of scaffolding work.
+1. No significant advantage over `.env` files placed with the code.
+
+Regardless of the storage mechanism we will need a few other key things.
+
+### Store
+
+To prevent feature flags from being all over the place, we can keep them store in a Pinia store and just use the store wherever there is a check. This allows us to toggle features from a centralised place.
+
+Since keys will be added and removed periodically, the store must not expose feature flags directly but rather expose a getter that can handle fallback-to-default.
```

Comment:
I'm not sure if the app should break down. I'm thinking of cases where links are shared with flag query params and visiting those links breaks the app after the feature flag has been turned on? There should be some mechanism of ignoring invalid flags or deprecated flags so that the app continues to work.

Reply:
Sorry, I meant a build-time breakage if an invalid feature flag name is passed to the checker.

Invalid query param feature flag names, I'm assuming, would just get thrown out. I can't see a situation in which the application would consume them and run them through the getter as it would imply that the frontend is for some reason taking the query param and during runtime, to decide whether a specific feature exists, using the query param. That doesn't make sense to me :thinking: 

To me a query param would manipulate the contents of the store (specifically allow setting switchable flags on or off) and the runtime code interacting with the store would just retrieve the current state.

In that scheme invalid query param flags would just be ignored. Maybe a helpful error in non-production environments could be shown to help quickly debug typos in flag names or something, like console logging "Found query param for feature flag `xyz` but could not find an actual matching flag with that name".

Reply:
Yeah, instead of crashing the app, we can just log warnings to the console whenever a non-existent flag is referenced in the code. Our intermediary layer for interacting with the flag should handle such mismatches.